### PR TITLE
Add KeysightE3631A

### PR DESCRIPTION
**Pull request by OptimisticBeliever**
Add the KeysightE3631A Triple DC Power Supply based on the keysightE36312A.
The test files with and without device are also available.

### CODE REVIEW

File: docs/api/instruments/keysight/keysightE3631A.rst [hunk 1]
```diff
@@ -0,0 +1,12 @@
+##############################################
+Keysight E36312A Triple Output Power Supply
+##############################################
```
Comment (bmoneke) [inline docs/api/instruments/keysight/keysightE3631A.rst:3]: Did you copy the text from the E3631**2**A?

You will have to cut the "bars" to the length of the text:

```suggestion
##########################################
Keysight E3631A Triple Output Power Supply
##########################################
```

File: pymeasure/instruments/keysight/keysightE3631A.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+    ch_3 = Instrument.ChannelCreator(VoltageChannel, 3)
+
+    def __init__(self, adapter, name="Keysight E3631A", **kwargs):
+        super().__init__(
+            adapter, name, **kwargs
+        )
+        self.channels[1].voltage_setpoint_values = [0, 6]
+        self.channels[1].current_limit_values = [0, 5]
+        self.channels[3].voltage_setpoint_values = [0, -25]
```
Comment (bmoneke) [inline pymeasure/instruments/keysight/keysightE3631A.py:91]: You could (it is a suggestion) do this work in the channel's `__init__` (by giving an appropriate parameter).

File: tests/instruments/keysight/test_keysightE3631A_with_device.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    @pytest.mark.parametrize("chn, voltage", [(1, 0), (1, 6), (2, 0), (2, 25), (3, 0), (3, -25)], )
+    def test_voltage_setpoint(self, instr, chn, voltage):
+        instr.ch(chn).voltage_setpoint = voltage
+        assert instr.ch(chn).voltage_setpoint == voltage
+
+    @pytest.mark.parametrize("chn, voltage", [(1, -1), (1, 7), (2, -1), (2, 26), (3, 1), (3, -26)], )
+    def test_voltage_setpoint_out_of_range(self, instr, chn, voltage):
+        with pytest.raises(ValueError) as not_in_range:
+            instr.ch(chn).voltage_setpoint = voltage
+            assert f"Value of {voltage} is not in range" in not_in_range
```
Comment (bmoneke) [inline tests/instruments/keysight/test_keysightE3631A_with_device.py:98]: There is a "match" parameter to check the error message, see the [pytest documentation](https://docs.pytest.org/en/4.6.x/reference.html#pytest-raises):
```suggestion
        with pytest.raises(ValueError, match=f"Value of {voltage} is not in range"):
            instr.ch(chn).voltage_setpoint = voltage
```

**Review by bmoneke — CHANGES_REQUESTED** [review]
Thank you for your contribution @OptimisticBeliever ,

you may add yourself to the `authors.txt` as you are a contributor with this contribution.

I added a few remarks, but your code looks already good.

File: pymeasure/instruments/keysight/keysightE3631A.py (hunk 2 above)
Comment (OptimisticBeliever) [reply pymeasure/instruments/keysight/keysightE3631A.py:91]: Hello @bmoneke and thanks for the quick replies.
I don't see the point here. By channel's `__init__` you mean add `def __init__` in the `class VoltageChannel(Channel)` ?

File: tests/instruments/keysight/test_keysightE3631A_with_device.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    This test suite, needs the following setup to work properly:
+        - A KeysightE3631A device should be connected to the computer;
+        - The device's address must be set in the RESOURCE constant;
+    """
+
+    ##################################################
+    # KeysightE3631A device address goes here:
+    RESOURCE = "GPIB0::10::INSTR"
+    ##################################################
```
[the remark this answers is not in the document]
Comment (OptimisticBeliever) [reply tests/instruments/keysight/test_keysightE3631A_with_device.py:43]: Thanks for the trick, but I don't know where/how this command line parameter has to be declared ? I'm using PyCharm IDE and I use to just hit the green arrow in front of the Class or test function to run the tests

[the remark this answers is not in the document]
Comment (OptimisticBeliever) [reply tests/instruments/keysight/test_keysightE3631A_with_device.py:43]: I add `--device-address "GPIB0::10::INSTR"` in the Additional Arguments of the Edit Run configuration and that works so I suppose this is the way to pass the command line parameter in PyCharm IDE.

[the remark this answers is not in the document]
Comment (bmoneke) [reply tests/instruments/keysight/test_keysightE3631A_with_device.py:43]: > I add `--device-address "GPIB0::10::INSTR"` in the Additional Arguments of the Edit Run configuration and that works so I suppose this is the way to pass the command line parameter in PyCharm IDE.

Normally you run the test files via a console command `pytest file_name`. There you can add that parameter. The IDE does that for you.

File: pymeasure/instruments/keysight/keysightE3631A.py (hunk 2 above)
Comment (OptimisticBeliever) [reply pymeasure/instruments/keysight/keysightE3631A.py:91]: Hello Benedikt,

I will stay as it is with the update of the dynamic properties `values `after the `class KeysightE3631A(Instrument)` init. Thanks for the suggestion.

**Comment by OptimisticBeliever** [comment]
Hello. May you review the last changes and proceed the missing workflow approval please ?

File: AUTHORS.txt [hunk 5]
```diff
@@ -65,4 +65,5 @@ Robert Roos
 Sebastien Weber
 Sebastian Neusch
 Ulrich Sauter
-Guus Kuiper
\ No newline at end of file
+Guus Kuiper
+Armindo PINTO
```
Comment (BenediktBurger) [inline AUTHORS.txt:69]: As you see from the other names, even the last name is in lower case (except for the first letter).
Unless your name is spelled that way officially, please adhere to the common standard.

Comment (OptimisticBeliever) [reply AUTHORS.txt]: Hello, yes sorry for the TYPO. Change has been committed.

File: pymeasure/instruments/keysight/keysightE3631A.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+        """Control the current limit of this channel, range depends on channel.""",
+        validator=strict_range,
+        values=[0, 1],
+        dynamic=True,
+    )
+
+    voltage = Channel.measurement(
+        "INST:NSEL {ch};:MEAS:VOLT?",
+        """Measure actual voltage of this channel.""",
+        dynamic=True
```
Comment (BenediktBurger) [inline pymeasure/instruments/keysight/keysightE3631A.py:59]: Why is this property dynamic? Copy paste error?
```suggestion
```

File: pymeasure/instruments/keysight/keysightE3631A.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+            adapter, name, **kwargs
+        )
+        self.channels[1].voltage_setpoint_values = [0, 6]
+        self.channels[1].current_limit_values = [0, 5]
+        self.channels[3].voltage_setpoint_values = [0, -25]
+
+    tracking_enabled = Instrument.control(
+        ":OUTP:TRAC?",
+        ":OUTP:TRAC %s",
+        """Control whether the power supply operate in the track mode (boolean)""",
```
Comment (BenediktBurger) [inline pymeasure/instruments/keysight/keysightE3631A.py:98]:
```suggestion
        """Control whether the power supply operates in the track mode (boolean)""",
```

File: pymeasure/instruments/keysight/keysightE3631A.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        validator=strict_discrete_set,
+        values={True: 1, False: 0},
+        map_values=True
```
Comment (BenediktBurger) [inline pymeasure/instruments/keysight/keysightE3631A.py:101]: It is good practice to terminate the last element with a comma as well (unless the parenthesis is on the same line).
That allows to add/remove more elements in the list/parameters, dict etc. without touching any other line due to adding/removing a comma.

I want to make you aware of that concept. We do not enforce it in pymeasure.

File: pymeasure/instruments/keysight/keysightE3631A.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        """
+        if channel_number == 1:
+            return self.ch_1
+        elif channel_number == 2:
+            return self.ch_2
+        elif channel_number == 3:
+            return self.ch_3
+        else:
+            raise ValueError("Invalid channel number. Must be 1, 2, 3.")
+
```
Comment (BenediktBurger) [inline pymeasure/instruments/keysight/keysightE3631A.py:130]: That method is superfluous as you can access the channels easily:
```python
inst.channels[1] == inst.ch_1
inst.channels[2] == inst.ch_2
inst.channels[3] == inst.ch_3
```

All three equalities are true as all Channels are added to the `channels` dictionary, where their `id` is the corresponding key.

File: pymeasure/instruments/keysight/keysightE3631A.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    ch_1 = Instrument.ChannelCreator(VoltageChannel, 1)
+
+    ch_2 = Instrument.ChannelCreator(VoltageChannel, 2)
+
+    ch_3 = Instrument.ChannelCreator(VoltageChannel, 3)
+
+    def __init__(self, adapter, name="Keysight E3631A", **kwargs):
+        super().__init__(
+            adapter, name, **kwargs
```
Comment (BenediktBurger) [inline pymeasure/instruments/keysight/keysightE3631A.py:89]: would you mind to add a parameter, whether the instrument uses SCPI or not (`includeSCPI`)?
Thanks.

File: pymeasure/instruments/keysight/keysightE3631A.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+        elif channel_number == 2:
+            return self.ch_2
+        elif channel_number == 3:
+            return self.ch_3
+        else:
+            raise ValueError("Invalid channel number. Must be 1, 2, 3.")
+
+    def reset(self):
+        """ Resets the instrument and clears the queue.  """
+        self.write("*RST;*CLS")
```
Comment (BenediktBurger) [inline pymeasure/instruments/keysight/keysightE3631A.py:133]: Attention: The instrument driver has already a `reset` method, which per default writes `*RST`. It also has another method (`clear`), which writes `*CLS`.

You could keep those two commands separated. Maybe you could rename that method to `reset_and_clear`, if you deem it necessary to have the two commands united in one.

**Review by BenediktBurger — CHANGES_REQUESTED** [review]
Thanks for your patience.

Here is the next round of comments (which caused two new issues regarding code improvement :grin:)

File: pymeasure/instruments/keysight/keysightE3631A.py (hunk 9 above)
Comment (OptimisticBeliever) [reply pymeasure/instruments/keysight/keysightE3631A.py:129]: Yes I know. This method is used in the test_keysightE3631A_with_device.py to dynamically pass the good channel in the test methods i.e `power_supply.ch(chn).voltage_setpoint == voltage`. I don't find another way to do that

File: pymeasure/instruments/keysight/keysightE3631A.py (hunk 10 above)
Comment (OptimisticBeliever) [reply pymeasure/instruments/keysight/keysightE3631A.py:88]: you mean this in the super init:
```suggestion
            adapter, name, includeSCPI = True, **kwargs
```

Comment (BenediktBurger) [reply pymeasure/instruments/keysight/keysightE3631A.py:88]: Exactly.
As it is a parameter, it should be `includeSCPI=True` (without spaces), though.

File: pymeasure/instruments/keysight/keysightE3631A.py (hunk 9 above)
Comment (BenediktBurger) [reply pymeasure/instruments/keysight/keysightE3631A.py:129]: what about `power_supply.channels[chn].voltage_setpoint == voltage`?

Comment (OptimisticBeliever) [reply pymeasure/instruments/keysight/keysightE3631A.py]: obviously...change has been made. Thank you.

File: pymeasure/instruments/keysight/keysightE3631A.py (hunk 10 above)
Comment (OptimisticBeliever) [reply pymeasure/instruments/keysight/keysightE3631A.py]: change committed.

File: pymeasure/instruments/keysight/keysightE3631A.py (hunk 11 above)
Comment (OptimisticBeliever) [reply pymeasure/instruments/keysight/keysightE3631A.py]: This method has been removed

**Comment by BenediktBurger** [comment]
May I merge?

**Comment by OptimisticBeliever** [comment]
Yes please. You can merge

**Pull request merged**